src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts

```diff
@@ -0,0 +1,209 @@
+		const language = await this._languageDetectionWorkerClient.detectLanguage(resource);
+		if (language) {
+			return this.getModeId(language);
+		}
+		return undefined;
+	}
+
+	async detectLanguages(resource: URI): Promise<string[]> {
+		const languages: Array<string | undefined> = await this._languageDetectionWorkerClient.detectLanguages(resource);
+		for (let i = 0; i < languages.length; i++) {
+			const modeId = this.getModeId(languages[i]);
+			languages[i] = modeId ? modeId : undefined;
+		}
+
+		return languages.filter(<T>(l?: T): l is T => Boolean(l));
+	}
+}
+
+export interface IWorkerClient<W> {
+	getProxyObject(): Promise<W>;
+	dispose(): void;
+}
+
+export class LanguageDetectionWorkerHost {
+	constructor(
+		private _indexJsUri: string,
+		private _modelJsonUri: string,
+		private _weightsUri: string,
+		private _telemetryService: ITelemetryService,
+	) {
+	}
+
+	async getIndexJsUri() {
+		return this._indexJsUri;
+	}
+
+	async getModelJsonUri() {
+		return this._modelJsonUri;
+	}
+
+	async getWeightsUri() {
+		return this._weightsUri;
+	}
+
+	async sendTelemetryEvent(languages: string[], confidences: number[], timeSpent: number): Promise<void> {
+		type LanguageDetectionStats = { languages: string; confidences: string; timeSpent: number; };
+		type LanguageDetectionStatsClassification = {
+			languages: { classification: 'SystemMetaData', purpose: 'FeatureInsight' };
+			confidences: { classification: 'SystemMetaData', purpose: 'FeatureInsight' };
+			timeSpent: { classification: 'SystemMetaData', purpose: 'FeatureInsight' };
+		};
+
+		this._telemetryService.publicLog2<LanguageDetectionStats, LanguageDetectionStatsClassification>('automaticlanguagedetection.stats', {
+			languages: languages.join(','),
+			confidences: confidences.join(','),
+			timeSpent
+		});
+	}
+}
+
+export class LanguageDetectionWorkerClient extends EditorWorkerClient {
+	private worker: IWorkerClient<LanguageDetectionSimpleWorker> | undefined;
+
+	constructor(
+		modelService: IModelService,
+		private readonly _telemetryService: ITelemetryService,
+		private readonly _indexJsUri: string,
+		private readonly _modelJsonUri: string,
+		private readonly _weightsUri: string) {
+		super(modelService, true, 'languageDetectionWorkerService');
+
+
+	}
+
+	private _getOrCreateLanguageDetectionWorker(): IWorkerClient<LanguageDetectionSimpleWorker> {
+		if (!this.worker) {
+
+			this.worker = this._register(new SimpleWorkerClient<LanguageDetectionSimpleWorker, EditorWorkerHost>(
+				this._workerFactory,
+				'vs/workbench/services/languageDetection/browser/languageDetectionSimpleWorker',
+				new EditorWorkerHost(this)
+			));
+		}
+		return this.worker;
+	}
+
+	override async _getProxy(): Promise<LanguageDetectionSimpleWorker> {
+		return await this._getOrCreateLanguageDetectionWorker().getProxyObject();
+	}
+
+	// foreign host request
+	public override fhr(method: string, args: any[]): Promise<any> {
```

Comment:
Why don't you make this method `async`, in that way you will not have to `Promise.resolve` wrap everything

Reply:
Funny enough, copilot wrote this entire message and I glanced and said "yeah that looks right"...async/await is better here I think.

Reply:
changed.